typhos/benchmark/profile.py

```diff
@@ -0,0 +1,120 @@
+"""
+Module using line_profiler to measure code performance and diagnose slowdowns.
+"""
+import importlib
+import pkgutil
+from inspect import isclass, isfunction
+
+from line_profiler import LineProfiler
+
+
+# Global profiler instance
+PROFILER=None
+
+
+def get_profiler():
+    """Returns the global profiler instance, creating it if necessary."""
+    global PROFILER
+    if PROFILER is None:
+        PROFILER = LineProfiler()
+    return PROFILER
+
+
+def setup_profiler(module_names=['typhos']):
+    """
+    Sets up the global profiler.
+
+    Includes all functions and classes from all submodules of the given
+    modules. This defaults to everything in the typhos module, but you can
+    limit the scope by passing a particular submodule,
+    e.g. module_names=['typhos.display']
+    """
+    functions = set()
+    for module_name in module_names:
+        modules = get_submodules(module_name)
+        for module in modules:
+            native_functions = get_native_functions(module)
+            functions.update(native_functions)
+
+    profiler = get_profiler()
+    for function in functions:
+        profiler.add_function(function)
+
+
+def toggle_profiler(turn_on):
+    """Turns the profiler off or on."""
+    profiler = get_profiler()
+    if turn_on:
+        profiler.enable_by_count()
+    else:
+        profiler.disable_by_count()
+
+
+def save_results(filename):
+    """Saves the formatted profiling results to filename."""
+    profiler = get_profiler()
+    with open(filename, 'w') as fd:
+        profiler.print_stats(fd)
+
+
+def print_results():
+    """Prints the formatted results directly to screen."""
+    profiler = get_profiler()
+    profiler.print_stats()
```

Comment:
My preferences for the output, but maybe config options would be nice.

* Remove items not called and switch to milliseconds as default units:
```suggestion
    profiler.print_stats(stripzeros=True, output_unit=1e-3)
```

Reply:
I think these make a lot of sense, I'll throw it in here and for the file dump